src/Control/Carrier/Choose/Church.hs

```diff
@@ -53,13 +53,18 @@ instance Fail.MonadFail m => Fail.MonadFail (ChooseC m) where
   fail s = lift (Fail.fail s)
   {-# INLINE fail #-}
 
+-- | Separate fixpoints are computed for each branch.
+--
+-- >>> run (runChooseS @[[Integer]] (pure . pure) (take 3 <$> mfix (\ as -> pure (0 : map succ as) <|> pure (0 : map pred as))))
+-- [[0,1,2],[0,-1,-2]]
 instance MonadFix m => MonadFix (ChooseC m) where
   mfix f = ChooseC $ \ fork leaf ->
-    mfix (runChoose
-      (liftA2 Fork)
-      (pure . Leaf)
-      . f . fromJust . fold (Alt.<|>) Just)
-    >>= fold fork leaf
+    mfix (runChooseS (pure . pure) . f . head)
+    >>= \case
+      a:|[] -> leaf a
+      a:|_  -> leaf a `fork` runChoose fork leaf (mfix (liftAll . fmap tail . runChooseS (pure . pure) . f))
+      where
+    liftAll m = ChooseC $ \ fork leaf -> m >>= foldr1 fork . fmap leaf
```

Comment:
Can you walk me through how this works? I’m not sure I understand where the `NonEmpty` is coming into play.

Reply:
`head` and `tail` are the versions from `Data.List.NonEmpty`, so

```haskell
(runChooseS (pure . pure) . f . head)
```

passes the head of the non-empty list to `f` and then eliminates the `ChooseC` in the result to a non-empty list, ultimately returned by `mfix`.

We then match on the list to see if there are more choices in its tail (= did `f` do any branching); if not, we return the sole result, otherwise we recur (i.e. call `mfix` on `ChooseC`, this time dropping the first branch.

This is the same approach taken by `ListT`. I’m not in love with calling `f` multiple times, but:

1. the need for `f` to control the evaluation of the fixed value means that I can’t see a way for us to avoid it; and
2. this is the only strategy I’m aware of that gives the desired semantics whereby each branch fixes its own value.

I think 1 is acceptable, because if you’re using `mfix` and nondeterminism, you have to expect that each branch is going to perform different effects; that’s what it means, really. And I think 2 is necessary, because otherwise `mfix` is rendered basically useless.

It’s also, as far as my reading of the `MonadFix` docs goes, lawful, so there’s that.